### ISSUE TITLE
Various errors from subtype.c

### ISSUE THREAD

**Issue by lkapelevich** [JuliaLang/julia#37685]
If I run [this gist](https://gist.github.com/lkapelevich/4101a1df2d4f6919b5871cd95e03fcb3) I get a trace that looks like [this](https://gist.github.com/lkapelevich/4101a1df2d4f6919b5871cd95e03fcb3#gistcomment-3462208), though I have seen some variation in the messages. 
The first commit I find that starts to trigger errors from the gist is https://github.com/JuliaLang/julia/commit/258d04e8bfdb8189fd61199bccd7641650fec07b. 

The first line I can catch in the referenced repo that causes the errors is [here](https://github.com/chriscoey/Hypatia.jl/blob/master/src/Cones/matrixepipersquare.jl#L378), it multiplies a general matrix by a Hermitian matrix. Unfortunately I haven't managed to get a more minimal example, or anything outside of the repo to trigger the errors. 
I'm using Julia from Ubuntu.

**Comment by lkapelevich** [comment]
Actually, I did get something more minimal:
```julia
using LinearAlgebra
mutable struct UntypedFact
    fact
    function UntypedFact()
        uf = new()
        return uf
    end
end

function foo(uf::UntypedFact)
    uf.fact = cholesky!(rand(1, 1))
    X = uf.fact \ rand(1, 1)
    Hermitian(X) * rand(1, 1)
    return
end

uf = UntypedFact()
foo(uf)

```

**Comment by maleadt** [comment]
Further reduced to:
```julia
import LinearAlgebra: Hermitian
function f(Type) Hermitian
end
function e(g)
    f(h)(fg)
end
function b(a)
    e(c) * Matrix{Float64}(undef, 1, 1)
end
d = 0
b(d)
```

Strangely I can't further reduce it by inlining the LinearAlgebra code, either just the type or the entire package, so maybe this is related to how the stdlib is compiled in the sysimg?

**Comment by chriscoey** [comment]
bump. the package Hypatia.jl passes tests on Travis with Julia 1.5, but fails due to this issue with Julia 1.6:
https://travis-ci.com/github/chriscoey/Hypatia.jl/jobs/395877922

**Comment by mtanneau** [comment]
I have the same issue [here](https://travis-ci.com/github/ds4dm/Tulip.jl/jobs/391607389) for Tulip.jl.

I'll try to dig into it as well; what puzzles me is that some of the functional tests run fine before the error occurs.

**Comment by vtjnash** [comment]
It's failing on:
```
using SparseArrays, LinearAlgebra
T = Tuple{typeof(mul!), StridedMatrix{T}, Hermitian{T, s809} where s809<:(StridedMatrix{T} where T), StridedMatrix{T}, Number, Number} where T<:Union{Float32, Float64}
S = Tuple{typeof(mul!), Matrix{T}, Hermitian{Tv, S} where S<:(Union{SparseArrays.AbstractSparseMatrixCSC{Tv, Ti}, SubArray{Tv, 2, s809, Tuple{Base.Slice{Base.OneTo{Int64}}, I}, L} where L where s809<:SparseArrays.AbstractSparseMatrixCSC{Tv, Ti} where I<:(AbstractUnitRange)} where Ti) where Tv, Matrix{Float64}, Int64, Int64} where T
typeintersect(T, S)
```

Which triggered subtyping to enter an infinite loop

**Comment by chriscoey** [comment]
@vtjnash thanks. this is a bug/regression that should be fixed before 1.6 is released, right?
